Detailed test plan for QoS Booking

jlurien opened the pull request:
#### What type of PR is this?

* enhancement/feature
* tests

#### What this PR does / why we need it:

Creates a new version of the test plan, substituting the single file that we had in the last meta, with a level of detail similar to what we have in other CAMARA APIs and aligned with the guidelines.

There is a separate file per operation, covering all expected responses. Comments are included when issues with the current spec are detected, which may be treated separately. 

#### Which issue(s) this PR fixes:

Fixes #69 

#### Special notes for reviewers:

Once merged, this test plan should evolve and be aligned with the API spec towards the next version.

#### Changelog input

```
- Enhanced test plan for QoS Booking API

```

jlurien committed:
Detailed test plan for QoS Booking

jlurien committed:
Merge branch 'main' into qos-booking-testplan-evolution

jlurien committed:
Linting error fixtures

jlurien committed:
Reduce long scenarios size

Masa8106 commented on code/Test_definitions/qos-booking-createBooking.feature:
```diff
@@ -0,0 +1,586 @@
+      | $.sink                   | exists only if provided in the request body and with the same value                                       |
+
+    # Open questions:
+    # serviceArea has to have same value as in the request body or implementations can make adjustments
```
Could you elaborate what you assume "implementations can make adjustments" is? Is there possibility that different serviceArea value with in the request body may be responded?

jlurien replied:
It's whether the value in the response has to be strictly the same as in the request or implementation may adjust the serviceArea to facilitate the implementation. I assume that it should be the same, but it would be good to make it more clear in the spec.

Masa8106 replied:
@jlurien, please let me suggest adding this proposal into line#50 and removing this note.
 
```suggestion
```

Masa8106 reviewed (commented):
Thank you for creating this PR.
Overall, it seems good. As there is several open questions, let us discuss them (Please find my comments.). Thank you.

Masa8106 commented on code/Test_definitions/qos-booking-createBooking.feature:
```diff
@@ -0,0 +1,586 @@
+    And the response property "$.code" is "QOS_BOOKING.AREA_NOT_COVERED"
+    And the response property "$.message" contains a user-friendly text
+
+## Code QOS_BOOKING.QOS_PROFILE_NOT_APPLICABLE to be added in next version
```
You are right. I missed it in QoD API. Thanks for clarification.

Masa8106 commented on code/Test_definitions/qos-booking-createBooking.feature:
```diff
@@ -0,0 +1,586 @@
+
+  # This step is optional, for cases when the implementation did not grant the booking synchronously and later rejects the booking
+  # or when it was granted (SCHEDULED) but implementation decides to revoke it before the start
+  # TBD if a better status or statusInfo is specified for this use case, e.g. status = "REJECTED" or status = "UNAVAILABLE" + statusInfo = "NOT_SCHEDULED" or "REVOKED"
```
Please refer to [Issue#83](https://github.com/camaraproject/QoSBooking/issues/83).
I would like to suggest removing this note.

```suggestion
```

Masa8106 commented on code/Test_definitions/qos-booking-createBooking.feature:
```diff
@@ -0,0 +1,586 @@
+    # startTime has to have same value as in the request body or implementations can make adjustments, to be returned when status = UNAVAILABLE?
+    # duration to be returned when when status = UNAVAILABLE?
+    # bookingId to be returned when when status = UNAVAILABLE?
+    # statusInfo should have another value when status = UNAVAILABLE after creation, e.g. BOOKING_CANNOT_BE_FULFILLED or open string
```
Please refer to Issue#83.
I would like to suggest removing this note.

```suggestion
```

Masa8106 commented on code/Test_definitions/qos-booking-createBooking.feature:
```diff
@@ -0,0 +1,586 @@
+    # duration to be returned when when status = UNAVAILABLE?
+    # bookingId to be returned when when status = UNAVAILABLE?
+    # statusInfo should have another value when status = UNAVAILABLE after creation, e.g. BOOKING_CANNOT_BE_FULFILLED or open string
+    # sinkCredential may be returned in the response in any case or it should be removed for security
```
I would like to suggest removing this note.

```suggestion
```

Masa8106 commented on code/Test_definitions/qos-booking-createBooking.feature:
```diff
@@ -0,0 +1,586 @@
+      | $.device                 | exists only if provided in the request body and contains only one of the identifier values in the request |
+      | $.qosProfile             | same value as in the request body                                                                         |
+      | $.startTime              | is in the future                                                                                          |
+      | $.status                 | is "REQUESTED", "SCHEDULED" or "UNAVAILABLE"                                                              |
```
I would like to suggest updating as below based on our latest discussion.

```suggestion
      | $.serviceArea        | same value as in the request body                                                              |
      | $.status                 | is "REQUESTED", "SCHEDULED" or "TERMINATED"                                                              |
```

jlurien committed:
Merge branch 'main' into qos-booking-testplan-evolution

jlurien committed:
Align test plan with PR #88 status model changes

Rename `status` property to `bookingStatus` and update status enum values
throughout all four feature files: AVAILABLE -> ACTIVATED, UNAVAILABLE ->
TERMINATED, and event data field `$.data.status` -> `$.data.bookingStatus`.

Resolve all open questions and TBDs:
- Add `$.serviceArea` equality assertion (agreed in CQM meeting 2026-02-06)
- Assert `$.sinkCredential` absent in all responses (security, aligned with QoD)
- Remove resolved open questions for startTime/duration/bookingId (required fields)
- Replace TBD on revoked booking with BOOKING_REVOKED statusInfo
- Add new scenario for synchronous booking decline (TERMINATED + BOOKING_DECLINED)

Fix deleteBooking async/sync split: 202 applies to REQUESTED/SCHEDULED/ACTIVATED;
204 applies to TERMINATED (not REQUESTED/UNAVAILABLE as previously stated).

Fix qos-booking-getBookingById scenario _02: replace non-existent property
`$.qosStatus` with correct `$.bookingStatus` = "TERMINATED".

Masa8106 reviewed (dismissed):
@jlurien, thanks for the update. LGTM.
Will you merge this PR? or I will do it tomorrow or onwards.
Thank you.

hdamker commented on code/Test_definitions/qos-booking-getBookingById.feature:
```diff
@@ -0,0 +1,170 @@
+    And the response property "$.applicationServerPorts" exists only if provided for createBooking and with the same value
+    And the response property "$.sink" exists only if provided for createBooking and with the same value
+    And the response property "$.sinkCredential" does not exist
+    And the response property "$.startedAt" exists only if "$.bookingStatus" is "ACTIVATED" and the value is in the past
```
Conflicts with YAML: code/API_definitions/qos-booking.yaml:438-439; terminated example includes startedAt at 1430-1447.

Masa8106 replied:
```suggestion
    And the response property "$.startedAt" exists only if "$.bookingStatus" is "ACTIVATED" and "TERMINATED", and the value is in the past
```
I will suggest the correction.

hdamker replied:
the bookingStatus can be only one of the options at a time:

```suggestion
    And the response property "$.startedAt" exists only if "$.bookingStatus" is "ACTIVATED" or "TERMINATED", and the value is in the past
```

hdamker commented on code/Test_definitions/qos-booking-createBooking.feature:
```diff
@@ -0,0 +1,603 @@
+    And the response header "x-correlator" has same value as the request header "x-correlator"
+    And the response header "Content-Type" is "application/json"
+    And the response property "$.status" is 422
+    And the response property "$.code" is "QOS_BOOKING.QOS_PROFILE_NOT_APPLICABLE"
```
Undefined error code QOS_BOOKING.QOS_PROFILE_NOT_APPLICABLE
Current 422 enum only includes defined codes at code/API_definitions/qos-booking.yaml:1318-1326.

Masa8106 replied:
I willl add the error code into qos-booking.yaml via another PR.

Masa8106 replied:
Have merged #104 .

hdamker commented on code/Test_definitions/qos-booking-deleteBooking.feature:
```diff
@@ -0,0 +1,197 @@
+    And the path parameter "bookingId" is set to the value for that QoS booking
+    When the request "deleteBooking" is sent
+    Then the response status code is 204
+    And the response header "Content-Type" is "application/json"
```
YAML 204 response has only x-correlator, no content: code/API_definitions/qos-booking.yaml:258-262.

Masa8106 replied:
```suggestion
```
I think we can remove text from this line.

hdamker commented on code/Test_definitions/qos-booking-createBooking.feature:
```diff
@@ -0,0 +1,603 @@
+    And the request body property  "$.sinkCredential.accessTokenExpiresUtc" is set to a value in the past
+    When the request "createBooking" is sent
+    Then the response status code is 400
+    And the response header "x-correlator" has same value as the request header "x-correlatofr"
```
```suggestion
    And the response header "x-correlator" has same value as the request header "x-correlator"
```

hdamker reviewed (commented):
A very few inconsistencies found by agent, see below.

Masa8106 committed:
Update code/Test_definitions/qos-booking-createBooking.feature

Co-authored-by: Herbert Damker <herbert.damker@telekom.de>

Masa8106 commented on code/Test_definitions/qos-booking-createBooking.feature:
```diff
@@ -0,0 +1,603 @@
+
+    Examples:
+      | property                 | condition                                                                                                 |
+      | $.device                 | exists only if provided in the request body and contains only one of the identifier values in the request |
```
As suggested in PR #97, I will suggest a change here.

```suggestion
      | $.device                 | exists only if provided in the request body and always if more than one device identifier is provided, And contains a single device identifier that was included in the request |
```

Masa8106 committed:
Update code/Test_definitions/qos-booking-createBooking.feature

jlurien committed:
Merge branch 'main' into qos-booking-testplan-evolution

jlurien commented on code/Test_definitions/qos-booking-createBooking.feature:
```diff
@@ -0,0 +1,603 @@
+    And the request body property  "$.sinkCredential.accessTokenExpiresUtc" is set to a value in the past
+    When the request "createBooking" is sent
+    Then the response status code is 400
+    And the response header "x-correlator" has same value as the request header "x-correlator"
```
```suggestion
    And the response header "x-correlator" has the same value as the request header "x-correlator"
```

jlurien committed:
Apply suggestions from code review

Co-authored-by: Masaharu Hattori <112887632+Masa8106@users.noreply.github.com>
Co-authored-by: Herbert Damker <herbert.damker@telekom.de>
Co-authored-by: Jose Luis Urien <jlurien@gmail.com>

Masa8106 reviewed (dismissed):
LGTM

hdamker reviewed (dismissed):
LGTM, see previous comment.

jlurien commented on code/Test_definitions/qos-booking-createBooking.feature:
```diff
@@ -0,0 +1,603 @@
+    And the response property "$.code" is "QOS_BOOKING.AREA_NOT_COVERED"
+    And the response property "$.message" contains a user-friendly text
+
+## Code QOS_BOOKING.QOS_PROFILE_NOT_APPLICABLE to be added in next version
```
```suggestion
```

jlurien committed:
Apply suggestion from @jlurien

jlurien commented on code/Test_definitions/qos-booking-createBooking.feature:
```diff
@@ -0,0 +1,602 @@
+    And the response property "$.code" is "QOS_BOOKING.AREA_NOT_COVERED"
+    And the response property "$.message" contains a user-friendly text
+
+
```
```suggestion
```

jlurien committed:
Apply suggestion from @jlurien

Masa8106 reviewed (approved):
I confirmed the removal of the comment. Thank you @hdamker and @jlurien .
Please merge this PR.
